From 03d96d365cd2fa68cb923f8ab95d03fd23b7c9f4 Mon Sep 17 00:00:00 2001
From: Min Fanlei <minfanlei@wxiat.com>
Date: Tue, 7 Jun 2022 17:41:51 +0800
Subject: [PATCH] sw64: add support for emulator running mode
Patch-mainline: NO, OTHERS
References: openEuler-22.03-LTS
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I56WV8

--------------------------------

We distinguish the running mode of emulator from guest by EMUL_FLAG
according to the related changes of qemu.

Signed-off-by: Min Fanlei <minfanlei@wxiat.com>

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Acked-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/chip/chip3/chip.c     |  2 +-
 arch/sw_64/include/asm/hw_init.h |  3 +++
 arch/sw_64/kernel/setup.c        | 14 +++++++++++---
 3 files changed, 15 insertions(+), 4 deletions(-)

diff --git a/arch/sw_64/chip/chip3/chip.c b/arch/sw_64/chip/chip3/chip.c
index adb4d325fc91..f4b17cfa1127 100644
--- a/arch/sw_64/chip/chip3/chip.c
+++ b/arch/sw_64/chip/chip3/chip.c
@@ -116,7 +116,7 @@ static int chip3_get_cpu_nums(void)
 
 static unsigned long chip3_get_vt_node_mem(int nodeid)
 {
-	return *(unsigned long *)MMSIZE;
+	return *(unsigned long *)MMSIZE & MMSIZE_MASK;
 }
 
 static unsigned long chip3_get_node_mem(int nodeid)
diff --git a/arch/sw_64/include/asm/hw_init.h b/arch/sw_64/include/asm/hw_init.h
index f3d75e0fda7a..ede8da6b0314 100644
--- a/arch/sw_64/include/asm/hw_init.h
+++ b/arch/sw_64/include/asm/hw_init.h
@@ -97,6 +97,9 @@ static inline bool icache_is_vivt_no_ictag(void)
 	return (cpu_desc.arch_var == 0x3 && cpu_desc.arch_rev == 0x1);
 }
 
+#define EMUL_FLAG	(0x1UL << 63)
+#define MMSIZE_MASK	(EMUL_FLAG - 1)
+
 DECLARE_STATIC_KEY_TRUE(run_mode_host_key);
 DECLARE_STATIC_KEY_FALSE(run_mode_guest_key);
 DECLARE_STATIC_KEY_FALSE(run_mode_emul_key);
diff --git a/arch/sw_64/kernel/setup.c b/arch/sw_64/kernel/setup.c
index 27f59fe0d1b1..e5abef2dc593 100644
--- a/arch/sw_64/kernel/setup.c
+++ b/arch/sw_64/kernel/setup.c
@@ -74,6 +74,7 @@
 #define DBGDCONT(args...)
 #endif
 
+
 DEFINE_PER_CPU(unsigned long, hard_node_id) = { 0 };
 
 #if defined(CONFIG_KVM) || defined(CONFIG_KVM_MODULE)
@@ -644,10 +645,17 @@ static void __init setup_cpu_info(void)
 	cpu_desc.va_bits = CPUID_VA_BITS(val);
 
 	if (*(unsigned long *)MMSIZE) {
-		pr_info("run mode: guest\n");
 		static_branch_disable(&run_mode_host_key);
-		static_branch_enable(&run_mode_guest_key);
-		static_branch_disable(&run_mode_emul_key);
+		if (*(unsigned long *)MMSIZE & EMUL_FLAG) {
+			pr_info("run mode: emul\n");
+			static_branch_disable(&run_mode_guest_key);
+			static_branch_enable(&run_mode_emul_key);
+
+		} else {
+			pr_info("run mode: guest\n");
+			static_branch_enable(&run_mode_guest_key);
+			static_branch_disable(&run_mode_emul_key);
+		}
 	} else {
 		pr_info("run mode: host\n");
 		static_branch_enable(&run_mode_host_key);
-- 
2.34.1

